File: services/spine/test/lib/telemetry_sampler_test.rb
require "test_helper"

class TelemetrySamplerTest < ActiveSupport::TestCase
  # -- background_chatter? -----------------------------------------------

  test "kube-probe user agent hitting /up" do
    assert TelemetrySampler.background_chatter?("http.user_agent" => "kube-probe/1.31", "http.target" => "/up")
  end

  test "elb-healthchecker user agent, mixed case" do
    # Regression guard: the Shuffler's sampler once lowercased the user agent and then
    # searched for "ELB-HealthChecker" (still mixed case) inside it, so it never matched.
    assert TelemetrySampler.background_chatter?("http.user_agent" => "ELB-HealthChecker/2.0", "http.target" => "/up")
  end

  test "the /up route, whatever is asking" do
    assert TelemetrySampler.background_chatter?("http.target" => "/up")
  end

  test "a probe user agent asking for anything at all" do
    assert TelemetrySampler.background_chatter?("http.user_agent" => "kube-probe/1.31", "http.target" => "/")
  end

  test "ordinary Spine traffic is not chatter" do
    ["/", "/admin/tables", "/tables/abc-123/events", "/tables/abc-123/seats"].each do |path|
      refute TelemetrySampler.background_chatter?("http.target" => path), "expected #{path} to not be chatter"
    end
  end

  test "a real browser hitting the admin screen" do
    refute TelemetrySampler.background_chatter?(
      "http.user_agent" => "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:153.0) Gecko/20100101 Firefox/153.0",
      "http.target" => "/admin/tables"
    )
  end

  test "a span with no HTTP attributes at all (an internal span)" do
    refute TelemetrySampler.background_chatter?({})
  end

  test "a cache-busting query string does not hide the path" do
    assert TelemetrySampler.background_chatter?("http.target" => "/up?check=1")
  end

  test "reads the stable url.path attribute too" do
    assert TelemetrySampler.background_chatter?("url.path" => "/up")
  end

  test "reads the stable user_agent.original attribute too" do
    assert TelemetrySampler.background_chatter?("user_agent.original" => "ELB-HealthChecker/2.0")
  end

  # -- BackgroundChatterSampler --------------------------------------------

  # A spread of trace ids across the low-64-bits id space that TraceIdRatioBased
  # actually compares against the ratio threshold (trace_id[8, 8] - the last 8 bytes).
  # A run of sequential/similar ids would all land on the same side and make the test
  # look like nothing (or everything) is kept, so scatter them: multiplying by Knuth's
  # constant twice pushes the product past 2**64 and lets the modulo wrap it around
  # several times, spreading values across the full range - deterministically, no
  # randomness, so the test isn't flaky.
  KNUTH = 2654435761
  TRACE_IDS = (1..2000).map do |i|
    ((i * KNUTH * KNUTH) % (2**64)).to_s(16).rjust(32, "0")
  end.freeze

  def sample_decision(sampler, attributes, trace_id)
    sampler.should_sample?(
      trace_id: [trace_id].pack("H*"),
      parent_context: OpenTelemetry::Context.empty,
      links: [],
      name: "GET",
      kind: :server,
      attributes: attributes
    )
  end

  def kept_count(attributes)
    sampler = TelemetrySampler::BackgroundChatterSampler.new
    TRACE_IDS.count { |id| sample_decision(sampler, attributes, id).sampled? }
  end

  test "records everything that is not chatter" do
    assert_equal TRACE_IDS.length, kept_count("http.target" => "/tables/abc-123/events")
  end

  test "drops the great majority of /up chatter" do
    assert_operator kept_count("http.target" => "/up"), :<, TRACE_IDS.length * 0.05
  end

  test "but keeps some chatter, so a failing probe is still visible" do
    assert_operator kept_count("http.target" => "/up"), :>, 0
  end
end

File: apps/tabletop/test/telemetry-sampler.test.ts
import { describe, test, expect } from "vitest";
import { SpanKind, ROOT_CONTEXT } from "@opentelemetry/api";
import { SamplingDecision } from "@opentelemetry/sdk-trace-base";
import {
  sampleRatioFor,
  BackgroundChatterSampler,
  KUBE_PROBE_SAMPLE_RATIO,
  ELB_HEALTHCHECKER_SAMPLE_RATIO,
  HEALTH_SAMPLE_RATIO,
  DEFAULT_SAMPLE_RATIO,
} from "../src/server/telemetry-sampler";

describe("sampleRatioFor", () => {
  test("kube-probe, whatever the path", () => {
    expect(sampleRatioFor({ "http.user_agent": "kube-probe/1.31", "http.target": "/" })).toBe(KUBE_PROBE_SAMPLE_RATIO);
    expect(sampleRatioFor({ "http.user_agent": "kube-probe/1.31", "http.target": "/health" })).toBe(KUBE_PROBE_SAMPLE_RATIO);
  });

  test("elb-healthchecker, whose user agent is mixed case", () => {
    expect(sampleRatioFor({ "http.user_agent": "ELB-HealthChecker/2.0", "http.target": "/health" })).toBe(ELB_HEALTHCHECKER_SAMPLE_RATIO);
  });

  test("the /health route, whatever is asking", () => {
    expect(sampleRatioFor({ "http.target": "/health" })).toBe(HEALTH_SAMPLE_RATIO);
  });

  test("a cache-busting query string does not hide the /health path", () => {
    expect(sampleRatioFor({ "http.target": "/health?v=3" })).toBe(HEALTH_SAMPLE_RATIO);
  });

  test("a normal route falls through to the default", () => {
    expect(sampleRatioFor({ "http.target": "/t/kitchen-table" })).toBe(DEFAULT_SAMPLE_RATIO);
  });

  test("a real browser hitting a table", () => {
    expect(
      sampleRatioFor({
        "http.user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:153.0) Gecko/20100101 Firefox/153.0",
        "http.target": "/t/kitchen-table",
      })
    ).toBe(DEFAULT_SAMPLE_RATIO);
  });

  test("a span with no HTTP attributes at all (an internal span)", () => {
    expect(sampleRatioFor({})).toBe(DEFAULT_SAMPLE_RATIO);
  });

  describe("newer OTel semantic conventions", () => {
    test("user_agent.original", () => {
      expect(sampleRatioFor({ "user_agent.original": "ELB-HealthChecker/2.0" })).toBe(ELB_HEALTHCHECKER_SAMPLE_RATIO);
    });

    test("url.path", () => {
      expect(sampleRatioFor({ "url.path": "/health" })).toBe(HEALTH_SAMPLE_RATIO);
    });
  });
});

describe("BackgroundChatterSampler", () => {
  const sampler = new BackgroundChatterSampler();

  function decide(attributes: Record<string, string>, traceId: string) {
    return sampler.shouldSample(ROOT_CONTEXT, traceId, "GET", SpanKind.SERVER, attributes, []).decision;
  }

  const traceIds = Array.from({ length: 2000 }, (_, i) => (((i + 1) * 2654435761) % 2 ** 32).toString(16).padStart(32, "0"));

  function keptCount(attributes: Record<string, string>) {
    return traceIds.filter(id => decide(attributes, id) === SamplingDecision.RECORD_AND_SAMPLED).length;
  }

  test("records everything on a normal route", () => {
    expect(keptCount({ "http.target": "/t/kitchen-table" })).toBe(traceIds.length);
  });

  test("drops the great majority of /health", () => {
    expect(keptCount({ "http.target": "/health" })).toBeLessThan(traceIds.length * 0.05);
  });

  test("but keeps some /health, so we can still see the health check succeeding", () => {
    expect(keptCount({ "http.target": "/health" })).toBeGreaterThan(0);
  });
});
